src/main/java/org/candlepin/pki/certs/X509CertificateBuilder.java

```diff
@@ -0,0 +1,307 @@
+
+        X509v3CertificateBuilder builder = new X509v3CertificateBuilder(
+            X500Name.getInstance(caCertificate.getSubjectX500Principal().getEncoded()),
+            this.certSerial,
+            Date.from(this.validAfter),
+            Date.from(this.validUntil),
+            new X500Name(this.distinguishedName.value()),
+            SubjectPublicKeyInfo.getInstance(clientPubKey.getEncoded()));
+
+        this.addSSLCertificateType(builder);
+        this.addKeyUsage(builder);
+        this.addAuthorityKeyIdentifier(builder, caCertificate);
+        this.addSubjectKeyIdentifier(builder, this.keyPair);
+        this.addSubjectAltName(builder, this.distinguishedName, this.subjectAltName);
+        this.addBasicConstraints(builder);
+        this.addExtensions(builder, this.certExtensions);
+
+        return buildCertificate(builder, this.signer());
+    }
+
+    private void addSSLCertificateType(X509v3CertificateBuilder builder) {
+        NetscapeCertType type = new NetscapeCertType(NetscapeCertType.sslClient | NetscapeCertType.smime);
+        addExtension(builder, MiscObjectIdentifiers.netscapeCertType, false, type);
+    }
+
+    private void addKeyUsage(X509v3CertificateBuilder builder) {
+        KeyUsage usage = new KeyUsage(KeyUsage.digitalSignature |
+            KeyUsage.keyEncipherment | KeyUsage.dataEncipherment);
+        this.addExtension(builder, Extension.keyUsage, false, usage);
+
+        ExtendedKeyUsage exUsage = new ExtendedKeyUsage(KeyPurposeId.id_kp_clientAuth);
+        this.addExtension(builder, Extension.extendedKeyUsage, false, exUsage);
+    }
+
+    private void addAuthorityKeyIdentifier(X509v3CertificateBuilder builder, X509Certificate caCertificate) {
+        try {
+            AuthorityKeyIdentifier aki = extensionUtil()
+                .createAuthorityKeyIdentifier(caCertificate);
+
+            builder.addExtension(Extension.authorityKeyIdentifier, false, aki.getEncoded());
+        }
+        catch (CertificateEncodingException | IOException e) {
+            throw new CertificateCreationException("Failed to add authority key identifier.", e);
+        }
+    }
+
+    private void addSubjectKeyIdentifier(X509v3CertificateBuilder builder, KeyPair keyPair) {
+        try {
+            byte[] ski = this.subjectKeyIdentifierWriter.getSubjectKeyIdentifier(keyPair);
+            builder.addExtension(Extension.subjectKeyIdentifier, false, ski);
+        }
+        catch (NoSuchAlgorithmException | IOException e) {
+            throw new CertificateCreationException("Failed to encode subject key identifier.", e);
+        }
+    }
+
+    private void addSubjectAltName(X509v3CertificateBuilder builder,
+        DistinguishedName distinguishedName, DistinguishedName subjectAltName) {
+
+        if (subjectAltName == null) {
+            return;
+        }
+
+        GeneralName subject = new GeneralName(GeneralName.directoryName, distinguishedName.value());
```

Comment:
Can we keep the implementation comment from BouncyCastlePKIUtility regarding subject alt names here? it provides details/rationale for why we're doing this, and referencing standards/RFC documents. I think this is the kind of knowledge we need to preserve